discovery_engine_core/core/src/storage/sqlite/setup.rs

```diff
@@ -0,0 +1,331 @@
+    if let Some(file_path) = &file_path {
+        let mut errors = vec![];
+
+        // if deletion fails it is often but not always a problem for the
+        // db recreation, so we log the error but do continue on.
+        remove_file_if_exists(file_path)
+            .await
+            .extract_error(&mut errors);
+        remove_file_if_exists(&format!("{}-wal", file_path))
+            .await
+            .extract_error(&mut errors);
+        remove_file_if_exists(&format!("{}-shm", file_path))
+            .await
+            .extract_error(&mut errors);
+
+        if errors.is_empty() {
+            Ok(())
+        } else {
+            Err(Error::Database(Box::new(CompoundError::new(
+                "deleting database files failed",
+                errors,
+            ))))
+        }
+    } else {
+        Ok(())
+    }
+}
+
+async fn init_storage_system_once(file_path: Option<String>) -> Result<SqliteStorage, Error> {
+    let pool = create_connection_pool(file_path.as_deref()).await?;
+    update_schema(&pool).await?;
+    update_static_data(&pool).await?;
+    Ok(SqliteStorage { file_path, pool })
+}
+
+pub(super) async fn create_connection_pool(file_path: Option<&str>) -> Result<Pool<Sqlite>, Error> {
+    let opt = if let Some(file_path) = &file_path {
+        SqliteConnectOptions::from_str(&format!("sqlite:{}", file_path))
+    } else {
+        SqliteConnectOptions::from_str("sqlite::memory:")
+    }?
+    .create_if_missing(true);
+
+    SqlitePoolOptions::new()
+        .connect_with(opt)
+        .await
+        .map_err(Into::into)
+}
+
+async fn update_schema(pool: &Pool<Sqlite>) -> Result<(), Error> {
+    sqlx::migrate!("src/storage/migrations")
+        .run(pool)
+        .await
+        .map_err(|err| Error::Database(err.into()))
+}
+
+//FIXME at some point we probably should derive this from the configuration
+//      and also decide what to do we documents in the history which are
+//      associated with a stack which is not included in the current config
+const EXISTING_STACKS: [stack::Id; 4] = [
+    stack::ops::breaking::BreakingNews::id(),
+    stack::ops::personalized::PersonalizedNews::id(),
+    stack::ops::trusted::TrustedNews::id(),
+    stack::exploration::Stack::id(),
+];
+
+async fn update_static_data(pool: &Pool<Sqlite>) -> Result<(), Error> {
+    let mut tx = pool.begin().await?;
+
+    let expected_ids = &EXISTING_STACKS;
+    let mut query_builder = QueryBuilder::new(String::new());
+    query_builder
+        .push("INSERT INTO Stack (stackId) ")
+        .push_values(expected_ids, |mut stm, id| {
+            stm.push_bind(id);
+        })
+        .push(" ON CONFLICT DO NOTHING;")
+        .build()
+        .persistent(false)
+        .execute(&mut tx)
+        .await?;
+    query_builder
+        .reset()
+        .push("DELETE FROM Stack WHERE stackId NOT IN ")
+        .push_tuple(expected_ids)
+        .build()
+        .persistent(false)
+        .execute(&mut tx)
+        .await?;
```

Comment:
:rocket:  If the array would be big enough then depending on how rustc/llvm would handle it it could make a relevant difference. But yes, here it's at best a pre-mature optimization probably not even that.